File: iosApp/iosApp/Pages/StopDetails/TileData.swift
//
//  TileData.swift
//  iosApp
//
//  Created by esimon on 11/27/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Foundation
import Shared

struct TileData: Identifiable {
    let id: String
    let route: Route
    let headsign: String
    let formatted: UpcomingFormat
    let upcoming: UpcomingTrip?

    init(
        route: Route,
        headsign: String,
        formatted: UpcomingFormat,
        upcoming: UpcomingTrip? = nil
    ) {
        self.route = route
        self.headsign = headsign
        self.formatted = formatted
        self.upcoming = upcoming
        id = upcoming?.trip.id ?? UUID().uuidString
    }

    init?(upcoming: UpcomingTrip, route: Route, now: Instant) {
        let formatted = if let formattedUpcomingTrip = RealtimePatterns.companion.formatUpcomingTrip(
            now: now,
            upcomingTrip: upcoming,
            routeType: route.type,
            context: .stopDetailsFiltered
        ) {
            UpcomingFormat.Some(trips: [formattedUpcomingTrip], secondaryAlert: nil)
        } else {
            UpcomingFormat.NoTrips(
                noTripsFormat: UpcomingFormat.NoTripsFormatPredictionsUnavailable()
            )
        }

        if !(formatted is UpcomingFormat.Some) {
            return nil
        }

        self.route = route
        headsign = upcoming.trip.headsign
        self.formatted = formatted
        self.upcoming = upcoming
        id = upcoming.trip.id
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/TileData.kt
package com.mbta.tid.mbta_app.android.stopDetails

import com.mbta.tid.mbta_app.model.RealtimePatterns
import com.mbta.tid.mbta_app.model.Route
import com.mbta.tid.mbta_app.model.TripInstantDisplay
import com.mbta.tid.mbta_app.model.UpcomingFormat
import com.mbta.tid.mbta_app.model.UpcomingTrip
import kotlinx.datetime.Instant

data class TileData(
    val id: String,
    val route: Route,
    val headsign: String,
    val formatted: UpcomingFormat,
    val upcoming: UpcomingTrip
) {
    companion object {
        fun fromUpcoming(upcoming: UpcomingTrip, route: Route, now: Instant): TileData? {
            val formattedUpcomingTrip =
                RealtimePatterns.formatUpcomingTrip(
                    now,
                    upcoming,
                    route.type,
                    context = TripInstantDisplay.Context.StopDetailsFiltered
                )
            val formatted =
                if (formattedUpcomingTrip != null) {
                    UpcomingFormat.Some(
                        trips = listOf(formattedUpcomingTrip),
                        secondaryAlert = null
                    )
                } else {
                    UpcomingFormat.NoTrips(
                        noTripsFormat = UpcomingFormat.NoTripsFormat.PredictionsUnavailable
                    )
                }

            if (formatted !is UpcomingFormat.Some) {
                return null
            }

            return TileData(upcoming.trip.id, route, upcoming.trip.headsign, formatted, upcoming)
        }
    }
}
